libs/http/include/http/connection.hpp

```diff
@@ -235,12 +241,15 @@ class HTTPConnection : public AbstractHTTPConnection,
   }
 
 private:
+  using RMutex = std::recursive_mutex;
+
   asio::ip::tcp::tcp::socket socket_;
   HTTPConnectionManager &    manager_;
   ResponseQueueType          write_queue_;
-  Mutex                      write_mutex_;
+  RMutex                     write_mutex_;
```

Comment:
This is the only logical change to the code (protect the is_open_, plus indicate it is not open on CloseConnection rather than Close.